marlowe/src/Language/Marlowe/Analysis/Safety/Ledger.hs

```diff
@@ -0,0 +1,318 @@
+  -> [SafetyError]  -- ^ The safety messages.
+checkTokens =
+  let
+    invalidToken token@(Token currency@CurrencySymbol{..} name@TokenName{..})
+      | currency == adaSymbol = if name == adaToken then mempty else pure $ InvalidToken token
+      | P.lengthOfByteString unCurrencySymbol /= 28 = pure $ InvalidCurrencySymbol currency
+      | P.lengthOfByteString unTokenName > 32 = pure $ TokenNameTooLong name
+      | otherwise = mempty
+  in
+   ((nub . foldMap invalidToken . toList) .) . extractAllWithContinuations
+
+
+-- | Check that a contract satisfies the maximum value ledger constraint.
+checkMaximumValueBound
+  :: Natural  -- ^ The `maxValueSize` protocol parameter.
+  -> Contract  -- ^ The contract.
+  -> Continuations  -- ^ The merkleized continuations.
+  -> [SafetyError]  -- ^ The safety messages.
+checkMaximumValueBound maxValueSize contract continuations =
+  let
+    actual = worstMaximumValue contract continuations
+  in
+    if actual <= maxValueSize
+      then mempty
+      else pure $ MaximumValueMayExceedProtocol actual
+
+
+-- | Compute a bound on the value size for a contract.
+worstMaximumValue
+  :: Contract  -- ^ The contract.
+  -> Continuations  -- ^ The merkleized continuations.
+  -> Natural  -- ^ A bound on the value size (in bytes).
+worstMaximumValue = (worstValueSize .) . extractAllWithContinuations
+
+
+-- | Find a representative value with worst-case size.
+worstValue
+  :: Contract  -- ^ The contract.
+  -> Continuations  -- ^ The merkleized continuations.
+  -> P.Value  -- ^ The value.
+worstValue contract continuations =
+  let
+    tokens = extractAllWithContinuations contract continuations
+    representativeValue (Token currency name) = V.singleton currency name big
+  in
+    V.singleton adaSymbol adaToken big
+      <> foldMap representativeValue tokens
+
+
+-- | Find a representative transaction output with worst-case size.
+worstTxOut
+  :: Contract  -- ^ The contract.
+  -> Continuations  -- ^ The merkleized continuations.
+  -> TxOut  -- ^ The transaction output.
+worstTxOut contract continuations =
+  TxOut
+  {
+    txOutAddress =
+      P.Address
+      {
+        addressCredential = PubKeyCredential "88888888888888888888888888888888888888888888888888888888"
+      , addressStakingCredential = Just . StakingHash $ PubKeyCredential "99999999999999999999999999999999999999999999999999999999"
+      }
+  , txOutValue = worstValue contract continuations
+  , txOutDatum = OutputDatumHash "5555555555555555555555555555555555555555555555555555555555555555"
+  , txOutReferenceScript = Nothing
+  }
+
+
+-- | Compute a bound on the minimum UTxO that might be required for a Marlowe contract.
+worstMinimumUtxo
+  :: Integer  -- ^ The `utxoCostPerByte` protocol parameter.
+  -> Contract  -- ^ The contract.
+  -> Continuations  -- ^ The merkleized continuations.
+  -> Integer  -- ^ A worst-case bound on the minimum UTxO lovelace for the contract.
+worstMinimumUtxo utxoCostPerByte contract continuations =
```

Comment:
Does this take into account any payments made from the deposit account holder during the contract? For example, if I use a 2 ADA minLovelace deposit but I subsequently pay 2 ADA from that account to a party, the min ADA deposit is no longer there.

Reply:
No, this is caught by the all-transactions analysis. I suppose that we could analyze the contract to see if it ever might withdraw the ada from the deposit account.